### PR TITLE
`Material.of(context)` in the test framework

### PR DESCRIPTION
**Pull request by nate-thegrate**
This PR renames the private `_RenderInkFeatures` class to `SplashController`, in preparation for an upcoming migration.
- [flutter.dev/go/layered-material-widgets](https://flutter.dev/go/layered-material-widgets)

<br><br>

Since **InkWell** handles ink splashes on its own, there's hardly ever a need for a developer to call `Material.of(context)` directly.

So let's use it to clean up some tests!

```dart
// before
final RenderObject inkFeatures = tester.allRenderObjects.firstWhere(
  (RenderObject object) => object.runtimeType.toString() == '_RenderInkFeatures',
);
expect(inkFeatures, paints..rect(color: focusColor));


// after
final BuildContext context = tester.element(find.text('button'));
expect(Material.of(context), paints..rect(color: focusColor));
```

<br>

**Edit:** I tried doing something fancy… **SplashController** is now an `extension type`! It exposes only the methods developers would want to see, and the type interface is "compiled away" completely and won't impact runtime performance :)

### CODE REVIEW

File: packages/flutter/lib/src/material/material.dart [hunk 1]
```diff
@@ -80,23 +80,19 @@ const Map<MaterialType, BorderRadius?> kMaterialEdges = <MaterialType, BorderRad
...
-  TickerProvider get vsync;
+//
+// TODO(nate-thegrate): deprecate this in the next PR!
+// https://github.com/flutter/flutter/pull/149963
+typedef MaterialInkController = SplashController;
 
+/// Allows a [MaterialInkController] to call the [addInkFeature] method.
+///
+/// Will be deprecated once the [Splash] class is implemented.
+extension AddInkFeature on MaterialInkController {
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/material.dart:91]: From the [style guide](https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#avoid-using-extension):

> ### Avoid using `extension`.
> Extension methods are confusing to document and discover. To an end developer, they appear no different than the built-in API of the class, and discovering the documentation and implementation of an extension is more challenging than for class members.
>
> Prefer instead adding methods directly to relevant classes. If that is not possible, create a method that clearly identifies what object(s) it works with and is part of.
>
> (A rare exception can be made for extensions that provide temporary workarounds when deprecating features. In those cases, however, the extensions and all their members must be deprecated in the PR that adds them, and they must be removed in accordance with our deprecation policy.)

In my mind, there are a few options:

1. Don't add this extension: instead, introduce SplashController with a deprecated `addInkFeature` method that points to `addSplash`
2. Don't change `addInkFeature` in this PR (wait to add the extension until the actual migration)
3. Add the extension now, so that SplashController can simply be picked up and moved to the widgets layer.

At this point I'm going with number 3 but am open to suggestions!

File: packages/flutter/lib/src/material/material.dart [hunk 2]
```diff
@@ -548,76 +543,77 @@ class _MaterialState extends State<Material> with TickerProviderStateMixin {
...
-  @override
+  /// (Optionally) specifies the color of the surface on which
+  /// splashes are being painted.
+  ///
+  /// If non-null, this [RenderBox] will absorb hit tests.
+  /// (The hit test behavior is usually inconsequential, since a splash
+  /// controller typically has an opaque parent, an opaque child, or both.)
   Color? color;
 
-  bool absorbHitTest;
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/material.dart:571]: Forgot to mention this in the PR description, but the redundant `absorbHitTest` flag is being removed as well.

File: packages/flutter/lib/src/material/material.dart [hunk 3]
```diff
@@ -684,8 +682,7 @@ abstract class InkFeature {
   ///
   /// Typically used by subclasses to call
   /// [MaterialInkController.markNeedsPaint] when they need to repaint.
-  MaterialInkController get controller => _controller;
-  final _RenderInkFeatures _controller;
+  final SplashController controller;
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/material.dart:685]: The "private member/public getter" pattern is no longer necessary!

File: packages/flutter/test/material/calendar_date_picker_test.dart [hunk 4]
```diff
@@ -1185,24 +1185,24 @@ void main() {
         theme: theme,
       ));
 
-      RenderObject inkFeatures = tester.allRenderObjects.firstWhere((RenderObject object) => object.runtimeType.toString() == '_RenderInkFeatures');
```
Comment (justinmc) [inline packages/flutter/test/material/calendar_date_picker_test.dart:1188]: It's nice to see `allRenderObjects` being removed here (and all over the place). I don't know if it will make a noticeable difference when running these tests but the new `Material.of` approach must be a lot more performant.

File: packages/flutter/test/material/navigation_bar_test.dart [hunk 5]
```diff
@@ -1145,11 +1145,13 @@ void main() {
     await gesture.moveTo(tester.getCenter(find.byType(NavigationIndicator).last));
     await tester.pumpAndSettle();
 
-    final RenderObject inkFeatures = tester.allRenderObjects.firstWhere((RenderObject object) => object.runtimeType.toString() == '_RenderInkFeatures');
+    // This test extracts the controller from the allRenderObjects iterable
+    // in order to prevent some wonkiness when running via skwasm.
+    final RenderObject controller = tester.allRenderObjects.firstWhere((RenderObject object) => object is SplashController);
```
Comment (justinmc) [inline packages/flutter/test/material/navigation_bar_test.dart:1150]: The `Material.of` approach was failing on skwasm?

File: packages/flutter/lib/src/material/material.dart [hunk 6]
```diff
@@ -548,76 +543,77 @@ class _MaterialState extends State<Material> with TickerProviderStateMixin {
...
-    RenderBox? child,
-    required this.vsync,
-    required this.absorbHitTest,
-    this.color,
-  }) : super(child);
+/// A [RenderBox] that paints [Splash]es.
+///
+/// Splash effects can be painted on an ancestor [Material] widget by obtaining
+/// an instance of this controller via [Material.of] and calling [addSplash].
+class SplashController extends RenderProxyBox {
```
Comment (justinmc) [inline packages/flutter/lib/src/material/material.dart:550]: Any benefit to moving this to its own file now that it's public?

File: packages/flutter/lib/src/material/material.dart (hunk 1 above)
Comment (justinmc) [reply packages/flutter/lib/src/material/material.dart:91]: With the current approach (3), there's some risk that a user will start depending on addInkFeature without seeing that it will be deprecated. Options 1 and 2 don't have this problem.

Actually can you explain the final state of addSplash and SplashController after all PRs are landed and deprecations removed? To make sure I'm seeing the big picture.

**Review by justinmc — COMMENTED** [review]
Looks good but I admit I never got around to reviewing https://github.com/flutter/flutter/pull/149963, so probably someone with more context from that PR should also review. Also still some discussion about the `extension`.

File: packages/flutter/lib/src/material/material.dart (hunk 1 above)
Comment (nate-thegrate) [reply packages/flutter/lib/src/material/material.dart:91]: Currently, we have an **_InkFeatures** widget that manages the **_RenderInkFeatures** render object.

**Material** is structured as a stateful widget, since it stores a global key for **_InkFeatures** and acts as a ticker provider for it as well.

Eventually, the goal is to have a **SplashBox** widget that acts as a minimal wrapper for **_InkFeatures** (or whatever we rename it to), and **Material** will be a stateless widget that uses **SplashBox** in its build method.

<br>

As far as **_RenderInkFeatures** is concerned, here's the plan:
- Change name to **SplashController**
- Remove the useless `absorbHitTest` flag
- Rename `addInkFeature` to `addSplash`

Functionally, there's no difference between keeping `addInkFeature` in the **SplashController** class and deprecating it, vs. converting to an extension method. But it might be nice for all the material-specific names to stay in the material library.

<br>

> there's some risk that a user will start depending on `addInkFeature` without seeing that it will be deprecated

I agree that there's some risk, but the only reason a developer would manually call `addInkFeature` is if they want to ignore `InkWell` entirely and implement their own "touch response" widget from scratch.
(If we'd like to deprecate `addInkFeature` in this PR, that would work too.)

File: packages/flutter/lib/src/material/material.dart (hunk 6 above)
Comment (nate-thegrate) [reply packages/flutter/lib/src/material/material.dart:550]: Unfortunately, I believe we'll have to wait until we migrate to **SplashBox**, since **_InkFeatures** needs access to **SplashController**'s private methods.

File: packages/flutter/test/material/calendar_date_picker_test.dart (hunk 4 above)
Comment (nate-thegrate) [reply packages/flutter/test/material/calendar_date_picker_test.dart:1188]: Agreed! I'm not 100% sure how the different `tester` methods compare, but if we see a sizeable performance boost that'd be awesome, and it's fun to make it more readable regardless :)

File: packages/flutter/test/material/navigation_bar_test.dart (hunk 5 above)
Comment (nate-thegrate) [reply packages/flutter/test/material/navigation_bar_test.dart:1150]: Yeah, it worked fine when I ran the test locally, but GitHub CI was not happy 😓

File: packages/flutter/lib/src/material/material.dart (hunk 1 above)
Comment (nate-thegrate) [reply packages/flutter/lib/src/material/material.dart:91]: Oh, I forgot to mention: currently, **MaterialInkController** is a public interface that is implemented by **_RenderInkFeatures**.

IMO the "public interface/private implementation" paradigm can be great in a lot of situations—in the case of the ink controller, I can see the appeal of not having all those render object methods in your autofill suggestions (the only alternative is to add `@protected` to all the methods you want to hide).
But since the interface is essentially not used at all, I feel that the only noticeable impact of this structure is how ugly it makes all those tests.

File: packages/flutter/lib/src/material/material.dart (hunk 6 above)
Comment (dkwingsmt) [inline packages/flutter/lib/src/material/material.dart:550]: It would puzzle me as a developer that a controller has methods of render boxes. Is the benefit mostly simplifying writing tests? I suppose the controller is used by end users a lot, and I prefer the previous way for a cleaner set of API.

File: packages/flutter/lib/src/material/material.dart [hunk 7]
```diff
@@ -548,76 +543,77 @@ class _MaterialState extends State<Material> with TickerProviderStateMixin {
...
+  /// [Splash]es added to this controller with [addSplash] use this vsync
+  /// to drive their animations.
   final TickerProvider vsync;
 
-  // This is here to satisfy the MaterialInkController contract.
-  // The actual painting of this color is done by a Container in the
-  // MaterialState build method.
-  @override
+  /// (Optionally) specifies the color of the surface on which
+  /// splashes are being painted.
```
Comment (dkwingsmt) [inline packages/flutter/lib/src/material/material.dart:561]: I don't remember seeing a lot of parentheses in Flutter docs so I suggest removing them, which reads fine too.
```suggestion
  /// Optionally specifies the color of the surface on which
  /// splashes are being painted.
```
The same for the one below.

File: packages/flutter/lib/src/material/material.dart [hunk 8]
```diff
@@ -548,76 +543,77 @@ class _MaterialState extends State<Material> with TickerProviderStateMixin {
...
+  ///
+  /// If non-null, this [RenderBox] will absorb hit tests.
+  /// (The hit test behavior is usually inconsequential, since a splash
+  /// controller typically has an opaque parent, an opaque child, or both.)
```
Comment (dkwingsmt) [inline packages/flutter/lib/src/material/material.dart:565]: What happens if the property is null? Will the splashes have no colors? How are the splashes useful then? (I mean it should be explained.)

File: packages/flutter/lib/src/material/material.dart (hunk 2 above)
Comment (dkwingsmt) [reply packages/flutter/lib/src/material/material.dart:571]: Can you illustrate how it is redundant? Is it straight up not used?

File: packages/flutter/lib/src/material/material.dart (hunk 3 above)
Comment (dkwingsmt) [reply packages/flutter/lib/src/material/material.dart:685]: I wonder what changed it? Was it forbidden to override `final` members?

**Review by dkwingsmt — COMMENTED** [review]
The change looks great and is a clean sub-PR from the big project! I have some suggestions for docs.

**Review by nate-thegrate — COMMENTED** [review]
Thank you @dkwingsmt for the great advice!

File: packages/flutter/lib/src/material/material.dart [hunk 9]
```diff
@@ -80,24 +80,12 @@ const Map<MaterialType, BorderRadius?> kMaterialEdges = <MaterialType, BorderRad
...
-  void addInkFeature(InkFeature feature);
-
-  /// Notifies the controller that one of its ink features needs to repaint.
-  void markNeedsPaint();
-}
+//
+// TODO(nate-thegrate): deprecate Material-specific symbols once Splash
+// fully replaces them.
+// The Material widget will be refactored to use a SplashBox in its build method.
+// https://flutter.dev/go/layered-material-widgets
```
Comment (dkwingsmt) [inline packages/flutter/lib/src/material/material.dart:87]: An issue will be preferred, since we can track the progress of the migration/project in the issue. You can link the doc in the issue.

File: packages/flutter/lib/src/material/material.dart (hunk 6 above)
Comment (dkwingsmt) [reply packages/flutter/lib/src/material/material.dart:542]: Thank you for explaining the context. But then why should we make it public (other than making tests simpler)? If the developers are not using it, we should just deprecate the public interface after renaming the private class.

Comment (nate-thegrate) [reply packages/flutter/lib/src/material/material.dart:542]: Good question, thank you.
I believe there are 2 reasons we'd want this to be public:

1. Since **InkWell** is defined in the material library, making **SplashController** private would prevent access.
2. A public class allows for custom **InkWell**/**SplashFactory** implementations.

<br>

I did a [GitHub-wide search](https://github.com/search?q=%22extends+InteractiveInkFeatureFactory%22++NOT+path%3A%22ink_sparkle.dart%22+NOT+path%3A%22ink_splash.dart%22+NOT+path%3A%22ink_ripple.dart%22+NOT+path%3A%22no_splash.dart%22+language%3ADart+&type=code&p=1) for **SplashFactory** subtypes: the vast majority of them had just copy-pasted existing logic with little or no changes. The same was true for [`Material.of(context)`](https://github.com/search?q=%22Material.of%28context%29%22++NOT+path%3A%22ink_well.dart%22+NOT+path%3A%22material.dart%22+NOT+path%3A%22ink_highlight.dart%22+NOT+path%3A%22ink_splash.dart%22+NOT+path%3A%22tabs.dart%22+NOT+path%3A%22text_field.dart%22+language%3ADart+&type=code).

Once we have a **SplashBox** API, perhaps more people will want to try making their own splash factories. But they would run into the exact same problems we have here: the current "public interface/private implementation" setup makes testing a whole lot more difficult.

So even in those cases where someone wants to make their own custom splash effects, I still believe that publicizing the implementation is an overall improvement for the developer experience.

Comment (dkwingsmt) [reply packages/flutter/lib/src/material/material.dart:542]: Thank you for the explanation. A few more questions:
* Should we name it in a more descriptive way? I mean, for a XxxController I'd expect it to be a mundane class, instead this is a render box. Why didn't we name it `RenderSplash`? (Actually I think I'm hoping we can signify its ability to hold multiple splash effects... but I think that's a topic we've discussed way back and I was probably convinced.)
* Also, have you experimented customizing splashes? Sometimes things that appear customizable don't end up  being so.

File: packages/flutter/lib/src/material/material.dart [hunk 10]
```diff
@@ -548,76 +535,89 @@ class _MaterialState extends State<Material> with TickerProviderStateMixin {
...
+  /// splashes are being painted.
+  ///
+  /// The splash controller doesn't paint this color—the value is stored here,
+  /// and descendants can access it via `Material.of(context).color`.
+  /// Unlike typical [InheritedWidget] `.of(context)` methods, no updates
+  /// are sent when the color changes.
+  ///
+  /// If non-null, this [RenderBox] will absorb hit tests.
+  /// (The hit test behavior is usually inconsequential, since a splash
+  /// controller typically has an opaque parent, an opaque child, or both.)
```
Comment (dkwingsmt) [inline packages/flutter/lib/src/material/material.dart:562]: I find this part of doc a bit hard to understand, and I think explaining one point per paragraph makes it clearer. I also adjusted wording a bit to emphasize the nature of being optional. What do you think of the following version?
```
  /// Suggests a color for the surface where splashes are painted.
  ///
  /// This color does not directly determine the appearance of the splash; 
  /// it serves as a reference value that descendants can optionally use 
  /// via `Material.of(context).color`. 
  ///
  /// Unlike standard [InheritedWidget] patterns, changes to this 
  /// color will **not** trigger automatic updates through `of(context)`.
  ///
  /// Additionally, if this color is non-null, the associated [RenderBox] 
  /// will absorb hit tests. However, this behavior usually has minimal impact 
  /// since splash controllers are often nested within widgets that are 
  /// either fully opaque, or have opaque children.
```

File: packages/flutter/lib/src/material/material.dart (hunk 6 above)
Comment (nate-thegrate) [reply packages/flutter/lib/src/material/material.dart:542]: > Why didn't we name it `RenderSplash`? (Actually I think I'm hoping we can signify its ability to hold multiple splash effects... but I guess that's a topic we've discussed ages ago and I was probably convinced back then.)

That's a really good question—the name `_RenderSplashes` would be analogous to the current `_RenderInkFeatures`.

This comment inspired a fun idea :)

<br>

> Also, have you experimented customizing splashes? Sometimes things that appear customizable don't end up being so.

Fortunately I don't think this will be a problem. The current `splashFactory` options are each defined in their own Dart files, which should hopefully show that a variety of effects can be painted without needing access to any private fields.

File: packages/flutter/lib/src/material/material.dart (hunk 10 above)
Comment (nate-thegrate) [reply packages/flutter/lib/src/material/material.dart:562]: This looks _much_ better, thanks!

**Comment by justinmc** [comment]
@nate-thegrate Heads up that you have a merge conflict.

File: packages/flutter/lib/src/material/material.dart [hunk 11]
```diff
@@ -638,37 +653,40 @@ class _InkFeatures extends SingleChildRenderObjectWidget {
...
+/// [SplashController] via [Material.of] and call [SplashController.addSplash].
 abstract class InkFeature {
   /// Initializes fields for subclasses.
   InkFeature({
-    required MaterialInkController controller,
+    required this.controller,
     required this.referenceBox,
     this.onRemoved,
-  }) : _controller = controller as _RenderInkFeatures {
+  }) {
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/material.dart:689]: A few advantages of the extension type:

- The type cast here is no longer required.
- Instead of only containing abstract fields, `SplashController` now houses the business logic for controlling splashes.
- An `abstract class` is often meant to be extended or implemented by multiple subtypes, whereas an `extension type`'s purpose is self-documenting.

**Comment by nate-thegrate** [comment]
Drafting until #158466 lands!

File: packages/flutter/lib/src/material/material.dart (hunk 11 above)
Comment (dkwingsmt) [reply packages/flutter/lib/src/material/material.dart:689]: If I understand correctly, this is also achievable via composition, i.e. making `_RenderInkFeatures` a private member of `SplashController`, and the extension type improves only on performance, since the compiler compiles them to the exact same type.

Comment (nate-thegrate) [reply packages/flutter/lib/src/material/material.dart:689]: Absolutely!

<br>

And aside from performance, there's one other benefit of being compiled to the same type:

```dart
bool doubleCheck(SplashController controller) {
  return controller is RenderBox; // Returns `true`, 100% guaranteed!
}
```

Since every splash controller is a `RenderBox` instance, the objects work beautifully with the [**PaintPattern**](https://main-api.flutter.dev/flutter/flutter_test/PaintPattern-class.html) API, allowing us to clean up all these tests:

```dart
final BuildContext context = tester.element(find.byType(DatePickerDialog));
final SplashController controller = Material.of(context);

expect(
  controller,
  paints..circle(color: datePickerTheme.dayOverlayColor?.resolve(const <WidgetState>{})),
);
```

We _could_ migrate the tests to `Material.of(context)` without the extension type, but the extension type provides improved type safety, as shown by the fact that `InkFeature` no longer needs any type casts!

Comment (dkwingsmt) [reply packages/flutter/lib/src/material/material.dart:689]: Hmm I don't quite get the type safety part you mentioned. I'm comparing the extension type with an approach that makes the `RenderBox` a member of `SplashController` (not with extending `RenderBox`). This approach also guarantees type safety. Querying the closest splash render box still needs some changes, but that won't be the biggest concern anyway.

Comment (nate-thegrate) [reply packages/flutter/lib/src/material/material.dart:689]: We currently have an `abstract class MaterialInkController` that's implemented by `_RenderInkFeatures`; this structure requires a type cast in the `InkFeature` constructor.

<br>

> an approach that makes the `RenderBox` a member of `SplashController` (not with extending `RenderBox`). This approach also guarantees type safety. Querying the closest splash render box still needs some changes, but that won't be the biggest concern anyway.

Yeah, that sounds right! Making `SplashController` a class with a private render object member would also ensure type safety:

```dart
class SplashController {
  SplashController._(this._renderBox);
  final _RenderInkFeatures _renderBox;

  // ...
}
```

<br>

However, this wouldn't allow us to clean up the tests:

```dart
testWidgets('SplashController test', (WidgetTester tester) async {
  final SplashController controller = Material.of(context);

  expect(
    controller,
    paints..circle(color: datePickerTheme.dayOverlayColor?.resolve(const <WidgetState>{})),
  );
});
```

```
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure was thrown running a test:
Expected: Object or closure painting: 'a circle'
  Actual: <Instance of 'SplashController'>
   Which: was not one of the supported objects for the "paints" matcher.
```

IMO the extension type is the clear winner here, in terms of its performance, clean interface, type safety, and testability.

Comment (dkwingsmt) [reply packages/flutter/lib/src/material/material.dart:689]: Can you explain what kind of cleanup you're looking for here?

So your PR produces:
```dart
    final BuildContext context = tester.element(find.byType(DatePickerDialog));
    final SplashController controller = Material.of(context);
    expect(
      controller,
      paints..circle(color: datePickerTheme.dayOverlayColor?.resolve(const <WidgetState>{})),
    );
```

It replaces the old `tester.allRenderObjects.firstWhere` with the internal name, which is a win for sure. But this is also achievable with my approach with minimal addition, such as exposing the object as a getter:
```dart
class SplashController {
  RenderObject get object => _object;
  RenderObject _object;
}
```
The test will look like
```dart
testWidgets('SplashController test', (WidgetTester tester) async {
  final SplashController controller = Material.of(context);

  expect(
    controller.object,
    paints..circle(color: datePickerTheme.dayOverlayColor?.resolve(const <WidgetState>{})),
  );
});
```

IMO the bigger problem is not how concise the test can be written, but more of the fact that a class structure we choose implies a relationship between the classes - just like the good old [is-a-vs-has-a case](https://www.w3resource.com/java-tutorial/inheritance-composition-relationship.php).

Making the controller and the object _the same_ object, whether through inheritance or extended class, implies that the controller _is_ a render object, which IMO is not correct. The splash controller _controls_ a group of splash effects, while a render object is a virtual representation of something drawn on the screen. While the splash effects will eventually be presented via an render object, it still is not the object itself. An extension class makes this situation worse, because it implies that "a splash controller is actually a render object in disguise", which is an incorrect way to think about it.

Comment (nate-thegrate) [reply packages/flutter/lib/src/material/material.dart:689]: You're absolutely right that exposing the render object as a getter would allow us to clean up the tests. The only downsides would be the class declaration incurring the performance cost of a wrapper, and that an extra public `RenderObject` field would show up in autofill suggestions.

Overall, I think we mostly share the same understanding, though I don't quite agree with a couple of things from the previous comment's final paragraph.

<br>

> a render object is a virtual representation of something drawn on the screen.

A [**RenderObject**](https://main-api.flutter.dev/flutter/rendering/RenderObject-class.html) is "an object in the render tree". In this PR, **SplashController** wraps the **_RenderSplashes** representation type, which extends off of [**RenderProxyBox**](https://main-api.flutter.dev/flutter/rendering/RenderProxyBox-class.html).

<br>

> An extension type makes this situation worse, because it implies that "a splash controller is actually a render object in disguise", which is an incorrect way to think about it.

I personally think that's a good way to think about it; maybe the ideal conceptualization would be "a render object with a custom interface".
Conceptually, it's very similar to the current [**MaterialInkController**](https://main-api.flutter.dev/flutter/material/MaterialInkController-class.html) class but has greater type safety. Rather than showing the **RenderProxyBox** inner workings, a **SplashController** exposes exactly what's needed so that descendant widgets can paint splashes.

Comment (dkwingsmt) [reply packages/flutter/lib/src/material/material.dart:689]: > The only downsides would be the class declaration incurring the performance cost of a wrapper

I'm sure this performance is negligible.

> and that an extra public RenderObject field would show up in autofill suggestions.

This is technically true, but I think the effect minimal since `SplashController` is not meant to be used by end developers much.

> I personally think that's a good way to think about it; maybe the ideal conceptualization would be "a render object with a custom interface".

If you really want to expose a render object with a custom interface, we can try, but we should not call it a controller. Flutter's other controllers are "simple" classes that are not exactly the tree node that they control, and it's a pattern I prefer. You can either expose a render object or expose a controller, but not expose a render object as the controller. 

And yes, we should expose limited interface to developers who want to control splashes, but again extension type isn't the only way, so this point does not prove that we need to do extension type. Wrapping with private property achieves everything you want, including limiting interface and improving type safety.

So basically using an extension type might have the following advantage over wrapping as a private property & getter:
* Performance gain (which is likely negligible)
* One fewer autofill suggestion

But has the following disadvantage:
* Incorrect usage of the _is-a_ model
* Require a new/special policy for extension type

Comment (nate-thegrate) [reply packages/flutter/lib/src/material/material.dart:689]: > If you really want to expose a render object (by itself) with a custom interface, we can do it as well, but we should call it something else instead of a controller.

That's a good point. We could call it **RenderSplashes** instead to match the private class name:

```dart
extension type RenderSplashes._(_RenderSplashes _renderObject) { ... }
```

I would be totally happy with this; the only disadvantage here would be a lack of support for data-driven fixes:
`final RenderSplashes controller` is a bit more confusing than `final SplashController controller`, and unfortunately we cannot rename variables with a `dart fix`.

The majority of Flutter's "controller" types are `Listenable`s; it'd be pretty awesome if we were 100% consistent and _every_ controller was a listenable. But if this is something we really did want to pursue, there'd be quite a bit of work to do:

```dart
// listenables                       // non-listenables
AnimationController                  MaterialInkController
CupertinoTabController               DrawerController
CarouselController                   ExpansionTileController
SearchController                     MenuController
TabController                        ScaffoldFeatureController
PlatformViewController               SystemContextMenuController
DraggableScrollableController        PersistentBottomSheetController
TextEditingController                ContextMenuController
TransformationController             HeroController
PageController                       MagnifierController
ScrollController                     OverlayPortalController
SnapshotController                   TreeSliverController
UndoHistoryController
WidgetStatesController
```

<br>

> advantages over wrapping as a private property & getter:
> 
> * Performance gain (which is likely negligible)
> * One fewer autofill suggestion
> 
> disadvantages:
> 
> * Incorrect usage of the _is-a_ model
> * Require a new/special policy for extension type

I think this is a really good summary, thank you.

And thanks for linking to the [is-a vs. has-a](https://www.w3resource.com/java-tutorial/inheritance-composition-relationship.php) page; it was well-written and very informative. It's super interesting to apply this distinction to extension types, since they satisfy _both_: an extension type that wraps a render object has a field pointing to that render object, and the extension type instance itself is a render object instance, according to type checks at runtime.

One could argue that we shouldn't ever use extension types, since they blur the lines of the is-a/has-a model. But I personally don't find that compelling, especially since these lines are already blurred elsewhere in the framework:

```dart
/// [AnimationController] is an [Animation<double>] instance…
class AnimationController extends Animation<double> {
  /// …and it has an [Animation<double>] field.
  Animation<double> get view => this;
}
```

(Though to be fair, I don't think the `.view` field is very useful and kinda wish we didn't have it.)

<br>

I completely agree that in this case, we don't _need_ to use an extension type, but I do feel that it's _better_ to do so.
In the case of **SplashController**, I see the is-a/has-a flexibility as a benefit, not a drawback; as far as an extension type policy, I think we'd be doing ourselves a disservice by not having one.

**Comment by nate-thegrate** [comment]
This is one I'm pretty bummed about closing, since it's a yak shave aimed to eventually solve [an issue](https://github.com/flutter/flutter/issues/147392) I filed 9 months ago.

Not the end of the world though!
